File: hooks/writing-prose-check.py
#!/usr/bin/env -S uv run python3
"""PostToolUse hook: prose-lint + structural constraints after draft edits.

Fires on Edit|Write to:
  - `drafts/*.md`   — markdown drafts (existing gate), AND
  - `*.typ` LETTERS — Typst letters (NOT slide decks; decks are skipped).

Two engines run, complementary and de-duplicated:
  1. scripts/prose-lint.py — the comprehensive PROSE PATTERN engine. Loads the
     regex tables defined once in the skill reference files (Strunk & White +
     ai-anti-patterns, plus econ/legal by domain) and lints the single edited
     file. Categories are domain-filtered via `--only` from
     `.planning/ACTIVE_WORKFLOW.md` `style:`.
  2. references/constraints/check-all.py — the GRANULAR engine, kept only for
     rules that need real logic, not regex tables (topic sentences, anchored
     numbers, outline sync, claim-id traceability, em-dash density, …).
     The three regex-table AI-smell constraints that DUPLICATE prose-lint's
     ai-anti-patterns category are suppressed here (see PROSE_LINT_SUPERSEDES)
     so the same rule isn't reported twice. check-all only globs `drafts/*.md`,
     so it runs for .md edits only; .typ letters get prose-lint alone.

Violations are filtered to the edited file AND the line range the edit actually
touched (±2 lines for boundary issues), so a single edit doesn't resurface
pre-existing warnings on unrelated lines.

Non-blocking: reports violations as an additionalContext message.
"""
from __future__ import annotations

import json
import re
import subprocess
import sys
from pathlib import Path

PLUGIN_ROOT = Path(__file__).parent.parent
CHECK_ALL = PLUGIN_ROOT / "references" / "constraints" / "check-all.py"
PROSE_LINT = PLUGIN_ROOT / "scripts" / "prose-lint.py"

# Granular constraints whose regex tables DUPLICATE prose-lint's
# `ai-anti-patterns` category. prose-lint is preferred; suppress these from the
# check-all results so the same rule is not double-reported. (em-dash density,
# topic sentences, anchored numbers, outline sync, etc. are real-logic checks
# with no prose-lint equivalent — they are NOT listed here and still run.)
PROSE_LINT_SUPERSEDES = {
    "writing-ai-smell-artifacts",
    "writing-ai-smell-puffery",
    "writing-ai-smell-structure",
}

# Domain `style:` -> extra prose-lint category. ai-anti-patterns and
# writing-general (Strunk) always run; econ/legal are added by domain.
_STYLE_CATEGORY = {
    "legal": "writing-legal",
    "econ": "writing-econ",
}

# Typst deck markers — if any appear we treat the .typ as a slide deck and skip.
# "touying"/"polylux" match the package import lines; "#slide(" matches a slide
# function call regardless of which deck framework defines it.
_DECK_MARKERS = ("touying", "polylux", "#slide(")
_DECK_DIR_RE = re.compile(r"^(slides|presentation)", re.IGNORECASE)


def _detect_style(project_root: Path) -> str | None:
    """Read `style:` from .planning/ACTIVE_WORKFLOW.md (legal/econ/general)."""
    aw = project_root / ".planning" / "ACTIVE_WORKFLOW.md"
    if not aw.is_file():
        return None
    try:
        m = re.search(r"^style:\s*(\w+)", aw.read_text(encoding="utf-8"),
                      re.MULTILINE)
        return m.group(1) if m else None
    except Exception:
        return None


def _prose_lint_categories(style: str | None) -> str:
    cats = ["ai-anti-patterns", "writing-general"]
    extra = _STYLE_CATEGORY.get((style or "").lower())
    if extra:
        cats.append(extra)
    return ",".join(cats)


def _is_typ_deck(path: Path) -> bool:
    """True if a .typ file is a slide deck (skip) rather than a letter."""
    for part in path.parts[:-1]:
        if _DECK_DIR_RE.match(part):
            return True
    try:
        text = path.read_text(encoding="utf-8", errors="ignore")
    except OSError:
        return False
    return any(marker in text for marker in _DECK_MARKERS)


def _edit_ranges(tool_name: str, tool_input: dict, path: Path) -> list[tuple[int, int]]:
    """Line ranges the edit touched. Write -> whole file; Edit -> new_string
    span ±2 lines; unlocatable -> whole file."""
    if tool_name == "Write":
        return [(1, 10**9)]
    new_string = tool_input.get("new_string", "")
    ranges: list[tuple[int, int]] = []
    if new_string and path.exists():
        try:
            file_text = path.read_text()
        except Exception:
            file_text = ""
        idx = file_text.find(new_string)
        while idx != -1:
            start_line = file_text.count("\n", 0, idx) + 1
            end_line = start_line + new_string.count("\n")
            ranges.append((max(1, start_line - 2), end_line + 2))
            idx = file_text.find(new_string, idx + 1)
    return ranges or [(1, 10**9)]


def _in_ranges(line_no: int, ranges: list[tuple[int, int]]) -> bool:
    return any(a <= line_no <= b for a, b in ranges)


def _run_prose_lint(path: Path, style: str | None,
                    ranges: list[tuple[int, int]]) -> list[str]:
    """Run prose-lint.py on the single edited file; return scoped violations."""
    try:
        proc = subprocess.run(
            [sys.executable, str(PROSE_LINT),
             "--only", _prose_lint_categories(style), str(path)],
            capture_output=True, text=True, timeout=30,
        )
    except Exception:
        return []
    # stdout lines pair up: locator line then an indented context line.
    # Locator format: `{path}:{line}:{col} [{category}] {label}`.
    loc_re = re.compile(rf"^{re.escape(str(path))}:(\d+):\d+\s+(\[.+)$")
    out: list[str] = []
    for line in proc.stdout.splitlines():
        m = loc_re.match(line)
        if not m:
            continue
        if _in_ranges(int(m.group(1)), ranges):
            out.append(f"{path.name}:{m.group(1)} {m.group(2)}")
    return out


def _run_check_all(project_root: Path, path: Path,
                   ranges: list[tuple[int, int]]) -> list[str]:
    """Run check-all.py; return logic/structural violations for the edited file,
    excluding constraints superseded by prose-lint."""
    try:
        proc = subprocess.run(
            [sys.executable, str(CHECK_ALL), str(project_root)],
            capture_output=True, text=True, timeout=30,
        )
        raw = proc.stdout.strip()
        brace_depth = 0
        json_end = 0
        for i, ch in enumerate(raw):
            if ch == "{":
                brace_depth += 1
            elif ch == "}":
                brace_depth -= 1
                if brace_depth == 0:
                    json_end = i + 1
                    break
        results = json.loads(raw[:json_end]) if json_end else {}
    except Exception:
        return []

    line_re = re.compile(rf"drafts/{re.escape(path.name)}:(\d+):")
    out: list[str] = []
    for entry in results.get("failed", []):
        name = entry.get("name", "")
        if name.rsplit("/", 1)[-1] in PROSE_LINT_SUPERSEDES:
            continue
        for v in entry.get("violations", []):
            m = line_re.search(v)
            if m and _in_ranges(int(m.group(1)), ranges):
                out.append(v)
    return out


def main():
    try:
        hook_input = json.load(sys.stdin)
    except Exception:
        sys.exit(0)

    tool_name = hook_input.get("tool_name", "")
    if tool_name not in ("Edit", "Write"):
        sys.exit(0)

    tool_input = hook_input.get("tool_input", {})
    file_path = tool_input.get("file_path", "")
    if not file_path:
        sys.exit(0)

    path = Path(file_path)
    suffix = path.suffix.lower()

    # Routing / scope gates.
    if suffix == ".md":
        if path.parent.name != "drafts":
            sys.exit(0)
        project_root = path.parent.parent
        run_check_all = True
    elif suffix == ".typ":
        if _is_typ_deck(path):
            sys.exit(0)
        project_root = path.parent
        run_check_all = False  # check-all only scans drafts/*.md
    else:
        sys.exit(0)

    style = _detect_style(project_root)
    ranges = _edit_ranges(tool_name, tool_input, path)

    violations = _run_prose_lint(path, style, ranges)
    if run_check_all:
        violations += _run_check_all(project_root, path, ranges)

    if not violations:
        sys.exit(0)

    output = ("Prose quality violations (scoped to edited lines):\n"
              + "\n".join(f"  • {v}" for v in violations))
    result = {
        "hookSpecificOutput": {
            "hookEventName": "PostToolUse",
            "additionalContext": output,
        }
    }
    print(json.dumps(result))
    sys.exit(0)


if __name__ == "__main__":
    main()

File: hooks/writing-prose-check.ts
#!/usr/bin/env bun
/**
 * PostToolUse hook: the deterministic prose audit + structural constraints after draft edits.
 *
 * Fires on Edit|Write to:
 *   - `drafts/*.md`   — markdown drafts (existing gate), AND
 *   - `*.typ` LETTERS — Typst letters (NOT slide decks; decks are skipped).
 *
 * ONE PROSE ENGINE, ONE STRUCTURAL ENGINE:
 *   1. scripts/prose-audit.py — every prose/AI-tell pattern system, de-duplicated, span-id'd.
 *   2. references/constraints/check-all.py — STRUCTURAL constraints only (bold-lead,
 *      topic-sentences, anchored-numbers, outline-sync): real logic, not regex over prose.
 *
 * WHY THAT SPLIT IS NOW A PREFIX RULE AND NOT A NAMED SET. This hook used to run prose-lint AND
 * check-all and suppress the overlap with `PROSE_LINT_SUPERSEDES`, a hand-maintained set of three
 * constraint names. It named the wrong three: the wikipedia-* tables are in BOTH engines and were
 * in neither's supersede list, so every AI-tell inside an edited range was reported to the model
 * twice. The set is gone. check-all's prose modules are skipped by the directory they live in,
 * which cannot fall out of date as tables are added. See
 * docs/DESIGN-prose-constraint-architecture.md.
 *
 * Neither engine is reimplemented here: both are the same Python scripts, spawned from the same
 * PLUGIN_ROOT-relative paths, so their exact strings and regex tables stay the source of truth.
 * Footnote masking, stylometrics and range-independent document signals now live inside
 * prose-audit.py rather than being assembled here.
 *
 * Non-blocking: reports violations as an additionalContext message.
 */
import { context, readPayload } from "./_gate_common";
import { existsSync, readFileSync } from "node:fs";
import { authenticatedWritingPlan } from "./lib/writing-plan-context.ts";
import { join, dirname } from "node:path";

const PLUGIN_ROOT = dirname(import.meta.dir);
const CHECK_ALL = join(PLUGIN_ROOT, "references", "constraints", "check-all.py");
const PROSE_AUDIT = join(PLUGIN_ROOT, "scripts", "prose-audit.py");

// Both scripts declare their own dependencies in a `uv run --with …` shebang; spawning them
// through the interpreter bypasses the shebang, so the deps are named here instead. Without
// lxml the .docx and wikipedia-* paths die at import; without pyyaml the diction tier does.
const PY = ["uv", "run", "--with", "lxml", "--with", "pyyaml", "python3"];

/** check-all constraint families that prose-audit.py already owns. A check-all entry name is
 *  `constraints/<stem>` or `skills/<skill>/references/<stem>`, so a directory prefix is enough. */
/*  `constraints/writing-no-bold-lead` is named explicitly rather than by directory: it is the one
 *  entry under `constraints/` that now DELEGATES to prose-audit.py (as `emphasis·bold-lead`), so
 *  leaving it in would report the same span from both engines — the exact double-reporting the
 *  prefix rule below exists to end. Its siblings under `constraints/` are structural and stay. */
const PROSE_ENGINE_PREFIXES = [
  "skills/ai-anti-patterns/",
  "skills/writing-",
  "constraints/writing-no-bold-lead",
];

const _DECK_MARKERS = ["touying", "polylux", "#slide("];
const _DECK_DIR_RE = /^(slides|presentation)/i;

// ---------------------------------------------------------------------------
// pathlib.Path semantics, only as far as this hook uses them.
// ---------------------------------------------------------------------------
function pyParts(p: string): string[] {
  const parts: string[] = [];
  if (p.startsWith("/")) parts.push("/");
  for (const seg of p.split("/")) {
    if (seg === "" || seg === ".") continue;
    parts.push(seg);
  }
  return parts;
}
function pyStr(p: string): string {
  const parts = pyParts(p);
  if (!parts.length) return ".";
  if (parts[0] === "/") return "/" + parts.slice(1).join("/");
  return parts.join("/");
}
function pyName(p: string): string {
  const parts = pyParts(p);
  const last = parts[parts.length - 1];
  return last === undefined || last === "/" ? "" : last;
}
function pyParent(p: string): string {
  const parts = pyParts(p);
  if (!parts.length) return ".";
  if (parts.length === 1) return parts[0] === "/" ? "/" : ".";
  const rest = parts.slice(0, -1);
  if (rest[0] === "/") return "/" + rest.slice(1).join("/");
  return rest.join("/");
}
function pySuffix(p: string): string {
  const name = pyName(p);
  if (name === "" || name === "." || name === "..") return "";
  const i = name.lastIndexOf(".");
  return i > 0 ? name.slice(i) : "";
}

function reEscape(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

// FOOTNOTE MASKING USED TO BE INLINED HERE, a hand port of scripts/lib/footnote_mask.py kept in
// sync by hand, because this hook fed style_metrics.py an on-disk draft directly. prose-audit.py
// masks all three formats itself (markdown, Typst, docx footnotes.xml) before any scorer sees the
// text, so the copy is gone rather than maintained in two languages.

type Range = [number, number];
const WHOLE_FILE: Range = [1, 10 ** 9];

function runPy(args: string[], cwd?: string): { stdout: string; ok: boolean } {
  try {
    const proc = Bun.spawnSync([...PY, ...args], {
      cwd,
      stdout: "pipe",
      stderr: "pipe",
      timeout: 30_000,
    });
    return { stdout: new TextDecoder().decode(proc.stdout), ok: true };
  } catch {
    return { stdout: "", ok: false };
  }
}

/** The plan's domain, normalized to prose-audit.py's `--style` vocabulary. Anything the audit
 *  does not know about degrades to `general` (Strunk + the AI-tell tables, no domain guide)
 *  rather than erroring out and leaving the draft unlinted. */
export function auditStyle(style: string | null): string {
  const s = (style || "").toLowerCase();
  return s === "legal" || s === "econ" ? s : "general";
}

export function isTypDeck(path: string): boolean {
  const parts = pyParts(path);
  for (const part of parts.slice(0, -1)) {
    if (_DECK_DIR_RE.test(part)) return true;
  }
  let text: string;
  try {
    text = readFileSync(path).toString("utf8"); // errors="ignore"-ish
  } catch {
    return false;
  }
  return _DECK_MARKERS.some((marker) => text.includes(marker));
}

export function editRanges(toolName: string, toolInput: Record<string, unknown>, path: string): Range[] {
  if (toolName === "Write") return [WHOLE_FILE];
  const newString = (toolInput.new_string as string) ?? "";
  const ranges: Range[] = [];
  if (newString && existsSync(path)) {
    let fileText = "";
    try {
      fileText = readFileSync(path, "utf8");
    } catch {
      fileText = "";
    }
    let idx = fileText.indexOf(newString);
    while (idx !== -1) {
      const startLine = countNewlines(fileText, 0, idx) + 1;
      const endLine = startLine + countNewlines(newString, 0, newString.length);
      ranges.push([Math.max(1, startLine - 2), endLine + 2]);
      idx = fileText.indexOf(newString, idx + 1);
    }
  }
  return ranges.length ? ranges : [WHOLE_FILE];
}

function countNewlines(s: string, start: number, end: number): number {
  let c = 0;
  for (let i = start; i < end; i++) if (s[i] === "\n") c++;
  return c;
}

export function inRanges(lineNo: number, ranges: Range[]): boolean {
  return ranges.some(([a, b]) => a <= lineNo && lineNo <= b);
}

/** THE prose engine. One span per violation, already collapsed across every pattern system, so
 *  what reaches the model is one line per finding rather than the same phrase under three labels. */
export function runProseAudit(path: string, style: string | null, ranges: Range[]): string[] {
  const { stdout, ok } = runPy([PROSE_AUDIT, "--json", "--style", auditStyle(style), path]);
  if (!ok) return [];
  let result: Record<string, unknown>;
  try {
    result = JSON.parse(stdout);
  } catch {
    return [];
  }
  const spans = (result.spans as Record<string, unknown>[]) ?? [];
  // A span at line 0 is document-level (diction saturation, per-section em-dash budget). It is
  // real, but it does not belong to any edited line, so it is only reported on a whole-file Write
  // — attaching it to an arbitrary Edit would repeat it on every subsequent keystroke.
  const wholeFile = ranges.length === 1 && ranges[0][0] === 1 && ranges[0][1] >= 10 ** 9;
  const out: string[] = [];
  for (const span of spans) {
    const line = Number(span.line ?? 0);
    if (line === 0 ? !wholeFile : !inRanges(line, ranges)) continue;
    // A span's labels carry their own `<system>: ` prefix so a reviewer citing a span id can tell
    // which table produced which label. In a one-line hook message the system is already the
    // bracket, so the prefix is stripped back off rather than printed twice.
    const labels = ((span.labels as string[]) ?? [])
      .map((l) => l.replace(/^[a-z0-9-]+: /, ""))
      .join(" | ");
    const where = line === 0 ? "doc" : String(line);
    const severity = String(span.severity ?? "soft");
    out.push(`${pyName(path)}:${where} [${span.system}${severity === "hard" ? "/HARD" : ""}] ${labels}`);
  }
  return out;
}

export function runCheckAll(projectRoot: string, path: string, ranges: Range[]): string[] {
  let results: Record<string, unknown> = {};
  try {
    const { stdout } = runPy([CHECK_ALL, projectRoot]);
    // Python iterates code POINTS and slices by them; mirror that so an astral char
    // ahead of the JSON blob cannot shift the cut.
    const raw = Array.from(stdout.trim());
    let braceDepth = 0;
    let jsonEnd = 0;
    for (let i = 0; i < raw.length; i++) {
      const ch = raw[i];
      if (ch === "{") braceDepth += 1;
      else if (ch === "}") {
        braceDepth -= 1;
        if (braceDepth === 0) {
          jsonEnd = i + 1;
          break;
        }
      }
    }
    results = jsonEnd ? JSON.parse(raw.slice(0, jsonEnd).join("")) : {};
  } catch {
    return [];
  }

  const lineRe = new RegExp(`drafts/${reEscape(pyName(path))}:(\\d+):`);
  const out: string[] = [];
  for (const entry of (results.failed as Record<string, unknown>[]) ?? []) {
    const name = (entry.name as string) ?? "";
    // Prose is prose-audit.py's job; check-all contributes only its structural constraints.
    if (PROSE_ENGINE_PREFIXES.some((p) => name.startsWith(p))) continue;
    for (const v of (entry.violations as string[]) ?? []) {
      const m = lineRe.exec(v);
      if (m && inRanges(parseInt(m[1], 10), ranges)) out.push(v);
    }
  }
  return out;
}

async function main(): Promise<void> {
  let hookInput: Record<string, unknown>;
  try {
    hookInput = await readPayload();
  } catch {
    process.exit(0);
  }

  const toolName = (hookInput.tool_name as string) ?? "";
  if (toolName !== "Edit" && toolName !== "Write") process.exit(0);

  const toolInput = (hookInput.tool_input as Record<string, unknown>) ?? {};
  const filePath = (toolInput.file_path as string) ?? "";
  if (!filePath) process.exit(0);

  const path = pyStr(filePath);
  const suffix = pySuffix(path).toLowerCase();

  let projectRoot: string;
  let runCheckAllFlag: boolean;
  if (suffix === ".md") {
    if (pyName(pyParent(path)) !== "drafts") process.exit(0);
    projectRoot = pyParent(pyParent(path));
    runCheckAllFlag = true;
  } else if (suffix === ".typ") {
    if (isTypDeck(path)) process.exit(0);
    projectRoot = pyParent(path);
    runCheckAllFlag = false; // check-all only scans drafts/*.md
  } else {
    process.exit(0);
  }

  const writingPlan = authenticatedWritingPlan(projectRoot);
  // Canonical writing hooks run only for an authenticated APPROVED receipt-selected plan.
  // Missing, pending, malformed, and legacy-only projects fail safe without lint output.
  if (!writingPlan) process.exit(0);
  const style = writingPlan.style || null;
  const ranges = editRanges(toolName, toolInput, path);

  let violations = runProseAudit(path, style, ranges);
  if (runCheckAllFlag) violations = violations.concat(runCheckAll(projectRoot, path, ranges));
  // Belt and braces on the de-duplication invariant: the audit collapses overlapping spans, and
  // check-all no longer contributes prose at all, so an identical line reaching here twice would
  // be a wiring regression. Report it once regardless.
  violations = [...new Set(violations)];

  if (!violations.length) process.exit(0);

  const output =
    "Prose quality violations (scoped to edited lines):\n" + violations.map((v) => `  • ${v}`).join("\n");
  context("PostToolUse", output);
}

// GUARDED SO THE MODULE CAN BE IMPORTED. Unguarded, `await main()` ran on import — reading stdin
// and exiting — so a test importing these helpers died before its first assertion and printed
// nothing, which reads as a pass. A hook has to be loadable to be unit-testable.
if (import.meta.main) await main();
